### PR TITLE
Linux micro kmod v4

### PR DESCRIPTION
**Pull request by dorileo**
## Changes
  + v2 (since #844)
    - impemented coldplug;
    - implemented hotplug;
    - fixed issues pointed in the last version;
    - introduced patches:
         - str-slice: implement trailing + leading whitespace helpers
         - file-reader: introduce sol_file_reader_fd() API
         - linux-micro: fix sysctl directory traversal

  + v3 (since #877)
      - fixed issues pointed by @lpereira and @barbieri in the last review (#877);

  + v4 (since #916)
     - fixed issues pointed by @lpereira in the last review (#916);
     - split the uevent handling function to its own source code (sharing between modules);
     - added the ```linux-micro: automount``` patch;
     - added unit tests to the patch ```str-slice: implement trailing + leading whitespace helpers```;

## Rationale
Implements the modprobe process following the systemd's modules-load.d
specification, also does the coldplug module loading - which consists in
looking up devices described in the sysfs and loading their respective
modules based on the modalias descriptors.

This module will also handle the hotplug events, notified by the kernel
via uevent netlink channel.

### CODE REVIEW

File: src/lib/datatypes/include/sol-str-slice.h [hunk 1]
```diff
@@ -123,6 +124,36 @@ sol_str_slice_to_string(const struct sol_str_slice slice)
...
+static inline struct sol_str_slice
+sol_str_slice_remove_trailing_whitespace(struct sol_str_slice slice)
+{
+    struct sol_str_slice copy = slice;
+    int len = copy.len;
+
+    if (copy.data[len] == '\0')
+        len--;
+
+    while (isspace(copy.data[len])) {
```
Comment (lpereira) [inline src/lib/datatypes/include/sol-str-slice.h]: Check if `len >= 0` before checking for `isspace(...)`.

File: src/lib/datatypes/include/sol-str-slice.h [hunk 2]
```diff
@@ -123,6 +124,36 @@ sol_str_slice_to_string(const struct sol_str_slice slice)
     return strndup(slice.data, slice.len);
 }
 
+static inline struct sol_str_slice
+sol_str_slice_remove_leading_whitespace(struct sol_str_slice slice)
+{
+    struct sol_str_slice copy = slice;
+
+    while (isspace(*copy.data) && copy.len) {
```
Comment (lpereira) [inline src/lib/datatypes/include/sol-str-slice.h]: Check `copy.len` before dereferencing `copy.data`.

File: src/modules/linux-micro/automount/automount.c [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+#include "sol-file-reader.h"
+#include "sol-mainloop.h"
+#include "sol-platform-linux-micro.h"
+#include "sol-uevent.h"
+#include "sol-util.h"
+
+#define EXTMAGIC "\123\357\001"
+#define EXT_SB_OFFSET 1024
+#define EXT_MAGIC_OFFSET EXT_SB_OFFSET + 0x38
```
Comment (lpereira) [inline src/modules/linux-micro/automount/automount.c]: This macro doesn't seem to be used.

File: src/test/test-str-slice.c [hunk 4]
```diff
@@ -108,6 +108,68 @@ test_str_slice_str_eq(void)
...
+
+#define TEST_EQUAL(X) { SOL_STR_SLICE_LITERAL(X), true }
+#define TEST_NOT_EQUAL(X) { SOL_STR_SLICE_LITERAL(X), false }
+
+    static const struct {
+        struct sol_str_slice input;
+        bool equal;
+    } table[] = {
+        TEST_NOT_EQUAL("with one trailing whitespace "),
+        TEST_NOT_EQUAL("with two trailing whitespace  "),
```
Comment (lpereira) [inline src/test/test-str-slice.c]: Might be interesting to test other kinds of whitespace that `isspace()` tests for, such as `\n` or `\t`.

**Comment by barbieri** [comment]
Mark some global structs/table as static.

Use sizeof() to declare magic whenever possible

Use memcmp() to check magic, if one day magic happens to include \0 the current code will fail.

I'm not sure we mount /mnt as tmpfs (see Linux micro main setup), if not add it there or use another mount point base. We shouldn't be touching the actual FS to create the mount point directories.

We need to find a way to request unmount. It shouldn't be Linux-micro specific, but we do need a way to explicitly get mount points and "eject" or unmount them. This way UI (or LED or webUI) can offer a way to save/load some stuff (i.e. Backup) and immediately unmount. Linux micro can hook to that request and implement it.

File: src/lib/datatypes/include/sol-str-slice.h [hunk 5]
```diff
@@ -123,6 +124,36 @@ sol_str_slice_to_string(const struct sol_str_slice slice)
...
+    }
+
+    return copy;
+}
+
+static inline struct sol_str_slice
+sol_str_slice_remove_trailing_whitespace(struct sol_str_slice slice)
+{
+    struct sol_str_slice copy = slice;
+    int len = copy.len;
```
Comment (lpereira) [inline src/lib/datatypes/include/sol-str-slice.h]: Why is `len` an `int`?

File: src/lib/datatypes/include/sol-str-slice.h [hunk 6]
```diff
@@ -123,6 +124,36 @@ sol_str_slice_to_string(const struct sol_str_slice slice)
...
+
+    if (copy.data[len] == '\0')
+        len--;
+
+    while (isspace(copy.data[len])) {
+        copy.len--;
```
Comment (lpereira) [inline src/lib/datatypes/include/sol-str-slice.h]: No need to update `copy.len` inside the loop, update it once right after the loop's over.

File: src/modules/linux-micro/automount/automount.c [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+
+    sb = (struct ext_super_block *)buff;
+    if (!(sb->s_feature_compat & EXT3_FEATURE_COMPAT_HAS_JOURNAL))
+        return false;
+
+    if ((sb->s_feature_ro_compat & EXT3_FEATURE_RO_COMPAT_UNSUPPORTED) ||
+        (sb->s_feature_incompat & EXT3_FEATURE_INCOMPAT_UNSUPPORTED))
+        return false;
+
+    return false;
```
Comment (lpereira) [inline src/modules/linux-micro/automount/automount.c]: All paths of this function returns `false`. Is it correct?

File: src/modules/linux-micro/automount/automount.c (hunk 3 above)
Comment (dorileo) [inline src/modules/linux-micro/automount/automount.c]: ok

File: src/lib/datatypes/include/sol-str-slice.h (hunk 5 above)
Comment (dorileo) [inline src/lib/datatypes/include/sol-str-slice.h]: cause the programmer is too stupid to not use ```size_t``` :) thanks...

File: src/lib/datatypes/include/sol-str-slice.h (hunk 6 above)
Comment (dorileo) [inline src/lib/datatypes/include/sol-str-slice.h]: actually it must be, ```len``` and ```copy.len``` may differ if dealing with null terminated string... probably we could change the whole approach...

File: src/modules/linux-micro/automount/automount.c (hunk 7 above)
Comment (dorileo) [inline src/modules/linux-micro/automount/automount.c]: yep, absolutely wrong, the last one should return ```true```.

File: src/lib/datatypes/include/sol-str-slice.h (hunk 1 above)
Comment (dorileo) [inline src/lib/datatypes/include/sol-str-slice.h]: ok

File: src/lib/datatypes/include/sol-str-slice.h (hunk 2 above)
Comment (dorileo) [inline src/lib/datatypes/include/sol-str-slice.h]: ouch, sure... thanks.

File: src/test/test-str-slice.c (hunk 4 above)
Comment (lpereira) [inline src/test/test-str-slice.c]: Also nice to test: declare the string `s = "  foo"`, and remove the leading whitespace from `s + 1`, to ensure that no buffer underflow will happen.

File: src/lib/datatypes/include/sol-str-slice.h (hunk 6 above)
Comment (dorileo) [inline src/lib/datatypes/include/sol-str-slice.h]: I changed the approach... to it doesn't apply anymore and don't face the address violation we saw with the sanitizer.

**Comment by dorileo** [comment]
v5 coming....
